### PR TITLE
Support logging received messages to a CSV file

### PR DESCRIPTION
**Pull request by rtzoeller**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/niveristand-ballard-arinc429-custom-device/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Add initial support for logging received messages to a CSV file.

Currently only the logging directory is configurable; eventually the format of the output file name as well as the format of the logged data will be configurable.

The output file currently looks like the following:

```
time stamp,channel,label,sdi,data word
0.305688,0,124,0,2326440060
0.305688,1,125,0,0178956413
0.306712,2,126,0,0178956414
0.306712,3,127,0,2326440063
0.306712,4,128,0,2326440064
0.307736,5,129,0,0178956417
0.307736,6,130,0,0178956418
0.308760,7,131,0,2326440067
0.405688,0,124,0,2326440060
```

### Why should this Pull Request be merged?

We want to eventually support logging all bus traffic. This is the first step.

### What testing has been done?

Deployed to a Linux RT controller and confirmed that looped-back traffic was logged to a file.

### CODE REVIEW

**Review by dbendele — COMMENTED** [review]
image

There is a weird coupling here in **Read from Hardware**. We lose the full data contained in the record, but we still build a very specific cluster, build an array of that indexing out of the For Loop, and then write that array of cluster to the **word log producer** class data. How would we enable verbose logging with the current implementation? One alternative might be to write array of records to class data (to enable maximum flexibility in _which_ columns get logged). Another alternative is to convert record to comma-separated string in the loop and send the array of strings to the logger (to provide greatest decoupling between custom device implementation and logging implementation.

**Comment by dbendele** [comment]
- [ ] Write to Hardware has no changes
- [ ] **Open Log File** and **Close Log File** are each no op VIs that have an error case structure that does the same nothing for **error in** = True or False. If there is a good reason to have the no-op VIs, let's remove the unnecessary case structure.
- [ ] **Write Logger** creates a DVR to a bundled class. What prevents creating a DVR to the class directly?
- [ ] **Open Log File** opens the log file with 'Create or Replace' option. As you mentioned earlier, we likely don't want to overwrite log files. Have you already updated the log file naming such that losing a log is unlikely/impossible?
- [ ] **Write Logged Data**
   - [ ] Recommend subVI to serialize **data**
   - [ ] Use fixed width (pad with zeros) for data word

**Comment by rtzoeller** [comment]
> There is a weird coupling here in Read from Hardware. We lose the full data contained in the record, but we still build a very specific cluster, build an array of that indexing out of the For Loop, and then write that array of cluster to the word log producer class data. How would we enable verbose logging with the current implementation? One alternative might be to write array of records to class data (to enable maximum flexibility in which columns get logged). How would we implement this for MIL-STD-1553 which supports word counts from 0 to 32? A different alternative could be to convert record to comma-separated string in the loop and send the array of strings to the logger (to provide greater decoupling between custom device implementation and logging implementation.

My top priority here was to minimize the impact on runtime performance. Logging should not impact the PCL rate, or should impact it to the minimum extent possible. To that effect, some amount of knowledge about what data is being logged is fundamentally required.

If we wanted to support verbose logging, the typedef of data we are bundling could easily be updated to include the extra fields.

I originally rejected the idea of writing it to class data because I wanted to minimize the number of allocations, and putting dynamic data in classes and building an array of those classes seemed fraught with peril.

Likewise, doing string formatting in the main thread seems like a bad idea, because they'll trigger allocations and are expensive. Doing that every time we read data is asking for trouble. IMO all string manipulation should happen on a separate thread, to the extent it is possible.

The idea that @buckd and I came up with is that the logging classes here would be code generated given a typedef for the data you want to log. For MIL-STD-1553, you would define a typedef of your data type which would include an array of words. This PR doesn't define a code generator, because I want to iterate on an instance of the logging classes before trying to programmatically generate them.

---

>  Open Log File and Close Log File are each no op VIs that have an error case structure that does the same nothing for error in = True or False. If there is a good reason to have the no-op VIs, let's remove the unnecessary case structure.

Are you referring to the VIs on `Word Logger Interface.lvclass`? The object is an interface, so we need the VIs, and conventionally interfaces provide empty implementations.

I can remove the case structures if you'd prefer, but the VIs and case structures were generated using LabVIEW's "New VI from Dynamic Dispatch Template". If the preference is to not have case structures in no-op VIs (which VIs on interfaces almost always should be), that seems like feedback for the LabVIEW team.

>  Write to Hardware has no changes

Good catch! Thanks.

> Write Logger creates a DVR to a bundled class. What prevents creating a DVR to the class directly?

LabVIEW doesn't allow this (no, seriously!).

From [here](https://www.ni.com/en-us/support/documentation/supplemental/09/reduce-memory-with-data-references.html#section--943600787): If you need a class reference that lets you change the underlying object, create a reference to a cluster of the class. This gives you the ability to swap values at the cost of hierarchical casting.

We are using an interface, so we fundamentally need to be able to swap the runtime type.

> Open Log File opens the log file with 'Create or Replace' option. As you mentioned earlier, we likely don't want to overwrite log files. Have you already updated the log file naming such that losing a log is unlikely/impossible?

No, my intent is to defer the configuration work out of this PR, and let this PR focus on the core logging behavior. I am okay with overwriting log files for now.

> Write Logged Data: Recommend subVI to serialize data

Makes sense. My intent is to provide some flexibility around formatting in a later PR, so that subVI can handle the various formatting options as well.

> Write Logged Data: Use fixed width (pad with zeros) for data word

Sure; I expect most people will opt for hex or octal instead of decimal anyways.

**Comment by dbendele** [comment]
> > There is a weird coupling here in Read from Hardware. We lose the full data contained in the record, but we still build a very specific cluster, build an array of that indexing out of the For Loop, and then write that array of cluster to the word log producer class data. How would we enable verbose logging with the current implementation? One alternative might be to write array of records to class data (to enable maximum flexibility in which columns get logged). How would we implement this for MIL-STD-1553 which supports word counts from 0 to 32? A different alternative could be to convert record to comma-separated string in the loop and send the array of strings to the logger (to provide greater decoupling between custom device implementation and logging implementation.
> 
> My top priority here was to minimize the impact on runtime performance. Logging should not impact the PCL rate, or should impact it to the minimum extent possible. To that effect, some amount of knowledge about what data is being logged is fundamentally required.
> 
> If we wanted to support verbose logging, the typedef of data we are bundling could easily be updated to include the extra fields.
> 
> I originally rejected the idea of writing it to class data because I wanted to minimize the number of allocations, and putting dynamic data in classes and building an array of those classes seemed fraught with peril.
> 
> Likewise, doing string formatting in the main thread seems like a bad idea, because they'll trigger allocations and are expensive. Doing that every time we read data is asking for trouble. IMO all string manipulation should happen on a separate thread, to the extent it is possible.
> 
> The idea that @buckd and I came up with is that the logging classes here would be code generated given a typedef for the data you want to log. For MIL-STD-1553, you would define a typedef of your data type which would include an array of words. This PR doesn't define a code generator, because I want to iterate on an instance of the logging classes before trying to programmatically generate them.
> 
> > Open Log File and Close Log File are each no op VIs that have an error case structure that does the same nothing for error in = True or False. If there is a good reason to have the no-op VIs, let's remove the unnecessary case structure.
> 
> Are you referring to the VIs on `Word Logger Interface.lvclass`? The object is an interface, so we need the VIs, and conventionally interfaces provide empty implementations.
> 
> I can remove the case structures if you'd prefer, but the VIs and case structures were generated using LabVIEW's "New VI from Dynamic Dispatch Template". If the preference is to not have case structures in no-op VIs (which VIs on interfaces almost always should be), that seems like feedback for the LabVIEW team.
> 
> > Write to Hardware has no changes
> 
> Good catch! Thanks.
> 
> > Write Logger creates a DVR to a bundled class. What prevents creating a DVR to the class directly?
> 
> LabVIEW doesn't allow this (no, seriously!).
> 
> From [here](https://www.ni.com/en-us/support/documentation/supplemental/09/reduce-memory-with-data-references.html#section--943600787): If you need a class reference that lets you change the underlying object, create a reference to a cluster of the class. This gives you the ability to swap values at the cost of hierarchical casting.
> 
> We are using an interface, so we fundamentally need to be able to swap the runtime type.
> 
> > Open Log File opens the log file with 'Create or Replace' option. As you mentioned earlier, we likely don't want to overwrite log files. Have you already updated the log file naming such that losing a log is unlikely/impossible?
> 
> No, my intent is to defer the configuration work out of this PR, and let this PR focus on the core logging behavior. I am okay with overwriting log files for now.
> 
> > Write Logged Data: Recommend subVI to serialize data
> 
> Makes sense. My intent is to provide some flexibility around formatting in a later PR, so that subVI can handle the various formatting options as well.
> 

Thanks for answering questions and addressing feedback!

> > Write Logged Data: Use fixed width (pad with zeros) for data word
> 
> Sure; I expect most people will opt for hex or octal instead of decimal anyways.

Regardless of the format, let's make sure that we don't lose real (transmitted/received) zeros from the end of the word.

**Review by buckd — COMMENTED** [review]
- [x] `Tx Scheduled Labels Execution Unit:Write to Hardware.vi` has no changes.
- [ ] `Add Logging Section.vi` output should not be **ports ref**.
image
- [ ] The `Logging` page's help points to the wrong item.
image
- [ ] I don't see any useful changes to `ActionVIOnCompile.vi`. Perhaps it is due to Class private data changes? Although, I would think source-only would prevent the need to resave this.

File: Source/Custom Device Support/Ballard ARINC 429 Support.lvproj [hunk 1]
```diff
@@ -968,6 +976,7 @@
 				<Property Name="Source[1].sourceInclusion" Type="Str">TopLevel</Property>
 				<Property Name="Source[1].type" Type="Str">Library</Property>
 				<Property Name="SourceCount" Type="Int">2</Property>
+				<Property Name="TgtF_enableDebugging" Type="Bool">true</Property>
```
[the remark this answers is not in the document]
Comment (rtzoeller) [reply Source/Custom Device Support/Ballard ARINC 429 Support.lvproj:979]: Unfortunately, yes.

[the remark this answers is not in the document]
Comment (rtzoeller) [reply Source/Custom Device Support/Ballard ARINC 429 Support.lvproj:979]: That said, I'm hopeful we can remove manually disable debugging from individual VIs which are in hot paths to minimize any performance impact.

**Review by buckd — APPROVED** [review]

**Pull request merged**